snapcraft/linters/metadata_linter.py

```diff
@@ -0,0 +1,193 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2022-2024 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Metadata linter implementation."""
+
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import Any
+
+import craft_cli
+from overrides import overrides
+
+from snapcraft.meta.snap_yaml import SnapMetadata
+
+from .base import Linter, LinterIssue, LinterResult
+
+_HELP_URL = "https://documentation.ubuntu.com/snapcraft/stable/reference/project-file/snapcraft-yaml/"
+
+
+@dataclass
+class MetadataField:
+    name: str
+    severity: LinterResult
+    extract: Callable[["SnapMetadata"], Any]
+    help_url: str
+    skip: Callable[["SnapMetadata"], bool] = lambda _: False
+
+
+def _get_links_attr(meta: SnapMetadata, key: str) -> list[str] | str | None:
+    """
+    Get the value of a links attribute.
+
+    :param meta: The snap metadata
+    :param key: The attribute name
+
+    :returns: The value of the attribute
+    """
+    if not meta.links:
+        return None
+
+    return getattr(meta.links, key, None)
+
+
+_FIELDS: list[MetadataField] = [
```

Comment:
@cmatsuoka, what do you think of the split here between some fields being a "warning" and others being "info"?

We're considering making them all the same level for simplicity.

Reply:
I know I wasn't asked, but I personally like the distinction between "you really should have this" and "maybe add this?"

Reply:
We're going to keep `info` and `warning` as-is.

Reply:
Thanks a lot for the update.